snap/local/build_and_install.sh

```diff
@@ -1,14 +1,50 @@
 #!/bin/bash
+# Cross-compile the Certbot snap from local sources for the specified architecture,
+# and install it if this architecture is also the the current machine one.
+# This script is designed for CI tests purpose.
+# Usage: build_and_install.sh [amd64,arm64,armhf]
 set -ex
 
-if [[ -z "$TRAVIS" ]]; then
+if [[ -z "${TRAVIS}" ]]; then
     echo "This script makes global changes to the system it is run on so should only be run in CI."
     exit 1
 fi
 
-sudo /snap/bin/lxd.migrate -yes
-sudo /snap/bin/lxd waitready
-sudo /snap/bin/lxd init --auto
-tools/strip_hashes.py letsencrypt-auto-source/pieces/dependency-requirements.txt > constraints.txt
-sudo snapcraft --use-lxd
-sudo snap install --dangerous --classic *.snap
+SNAP_ARCH=$1
+
+if [[ -z "${SNAP_ARCH}" ]]; then
+    echo "You need to specify the target architecture"
+    exit 1
+fi
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+CERTBOT_DIR="$(dirname "$(dirname "${DIR}")")"
+
+# shellcheck source=common.sh
+source "${DIR}/common.sh"
+
+docker run --rm --privileged multiarch/qemu-user-static:register --reset
+QEMU_ARCH=$(GetQemuArch "${SNAP_ARCH}")
+DOCKER_ARCH=$(GetDockerArch "${SNAP_ARCH}")
+
+docker run --net=host -d --rm -v "${DIR}/packages:/data/packages" --name pypiserver pypiserver/pypiserver
```

Comment:
I thought the PyPI server would add some sort of indexing in order to be able to retrieve efficiently the list of available/compatible wheels, but if it does not work that way, I can try it then.

The only drawback I see is that it would add Python as system dependency to build snaps.

Reply:
Yeah I tested something like this locally and it worked. We also do similar things in our [test farm tests](https://github.com/certbot/certbot/blob/8c8d3fab919a98cf0913ba8dff681ff75cc3178d/tests/letstest/scripts/test_leauto_upgrades.sh#L48) and [release script](https://github.com/certbot/certbot/blob/8c8d3fab919a98cf0913ba8dff681ff75cc3178d/tools/_release.sh#L160).

I think we already have the Python dependency with things like `tools/strip_hashes.py`.

Reply:
Ok, approach already used + no more dependencies, it works for me.

Reply:
I tried the approach with a simple HTTP server. In fact I am not that convinced after doing the actual implementation. It is true that the dockerized PyPI server is overkill in term of functionalities, but it is still a simple call to Docker, which is equivalent in term of complexity to call `tools/simple_http_server.py`.

However it adds a significant complexity to `snap/local/compile_native_wheels` by adding extra directory creation, copy and chmod, because for the plain HTTP server, wheels needs to be in directory named after the package names, while the PyPI server just needs a plain directory with all the wheels in it, and will dynamically expose the correct PyPI API. Of course we could abstract the creation of each package class as you proposed, but at this point it is also overkill for only two packages.

For these reasons, and since Docker is a runtime dependency anyway here, I prefer to stay with the Docker approach to serve the wheels.

Reply:
I'd still personally like to see the change I proposed here.

My concern isn't with Docker, but adding a dependency on `pypiserver/pypiserver`. I'd bet the project and its maintainers are great and I don't have any specific concerns, but it is an only moderately popular 3rd party reimplementation of PyPI. Will their Docker images ever break or disappear? Will they ever change the image's behavior in a way that breaks things for us? I don't know.

I do know that things like `mkdir`, `cp`, `chmod` and many popular HTTP servers are extraordinarily stable though and we can use them to remove our dependency on `pypiserver/pypiserver` in at most a few lines of additional code. Because of that, I personally still think it's worth it to do that work to remove the dependency and potential issues that come with every new dependency we add.